### PR TITLE
Add .NET CLR build script

### PR DESCRIPTION
**Pull request by omern1**
For https://github.com/compiler-explorer/compiler-explorer/pull/2592.

### CODE REVIEW

File: build/buildCLR.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+set -exu
+
+RUNTIME_VERSION="6.0.0-preview.5.21301.5"
+
+
+git clone --single-branch --depth 1 -b v$RUNTIME_VERSION "https://github.com/dotnet/runtime.git"
+cd runtime
+
+./build.sh clr --clang12 --checked
```
Comment (mattgodbolt) [inline build/buildCLR.sh:11]: We'll need to output the results somewhere. If you look at the other build scripts (e.g. buildcc65) they're responsible for archiving their output and then copying that tarball somewhere (often S3), What artifacts do we need to save from this build?

**Review by mattgodbolt — COMMENTED** [review]
I just tried this locally:

```
+ ./build.sh clr --clang12 --checked
__DistroRid: linux-x64
Downloading 'https://dot.net/v1/dotnet-install.sh'
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.

dotnet-install: Downloading primary link https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.3.21202.5/dotnet-sdk-6.0.100-preview.3.21202.5-linux-x64.tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-preview.3.21202.5/dotnet-sdk-6.0.100-preview.3.21202.5-linux-x64.tar.gz
dotnet-install: Adding to current process PATH: `/root/runtime/.dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Note that the script does not resolve dependencies during installation.
dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install, select your operating system and check the "Dependencies" section.
dotnet-install: Installation finished successfully.

Welcome to .NET 6.0!
---------------------
SDK Version: 6.0.100-preview.3.21202.5

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
MSBUILD : error MSB1001: Unknown switch.
Switch: --checked

For switch syntax, type "MSBuild -help"
Build failed with exit code 1. Check errors above.
```

**Comment by chausner** [comment]
> I just tried this locally:

There is no "--checked" option but there is `--configuration Checked` and `--runtimeConfiguration Checked`, see https://github.com/dotnet/runtime/blob/main/eng/build.sh#L24 and https://github.com/dotnet/runtime/blob/main/eng/build.sh#L35. I am not quite sure what the difference between the two is.

File: build/buildCLR.sh (hunk 1 above)
Comment (chausner) [reply build/buildCLR.sh:11]: The primary build output is created in `./artifacts/bin/coreclr/Linux.x64.Checked`, see https://github.com/compiler-explorer/compiler-explorer/blob/b78dfa6e24685e327ec3d762e04e671bf60d0d46/etc/config/csharp.defaults.properties#L13.

Comment (chausner) [reply build/buildCLR.sh:11]: Also `./.dotnet` seems to be required, see https://github.com/compiler-explorer/compiler-explorer/blob/b78dfa6e24685e327ec3d762e04e671bf60d0d46/etc/config/csharp.defaults.properties#L10.

**Comment by chausner** [comment]
I was able to build the runtime successfully on Ubuntu 20.04 using `./build.sh clr --clang12` after installing these prerequisites:
- apt packages: libkrb5-dev libssl-dev libicu-dev liblttng-ust-dev
- Clang12 installed in PATH

File: build/buildCLR.sh (hunk 1 above)
Comment (chausner) [reply build/buildCLR.sh:11]: Both `./artifacts/bin/coreclr/Linux.x64.Checked` and `./.dotnet` are created by `build.sh`, so I guess you can just tar gzip both folders and upload them to S3.

**Comment by mattgodbolt** [comment]
Cool! Taking a look now

File: build/buildCLR.sh (hunk 1 above)
Comment (mattgodbolt) [reply build/buildCLR.sh:11]: @chausner - the idea is that if you're submitting a PR to add this, we'd hope you could do this. There are many scripts to copy from to do this. All the other build.sh scripts have a few lines at the end to copy files around.

**Comment by mattgodbolt** [comment]
Hi.  This still fails with the same error:

```
--------------------------------------------------------------------------------------
MSBUILD : error MSB1001: Unknown switch.
Switch: --checked

For switch syntax, type "MSBuild -help"
Build failed with exit code 1. Check errors above.
root@20c87a642d19:~# 
```

I am testing this by:

* buliding the docker image (`docker build . --tag gcc-misc-builder:latest`)
* running the `buildCLR.sh` that you provided here in this PR inside the docker container: `docker run --rm -it gcc-misc-builder:latest bash buildCLR.sh`

By running everything in the docker image that means we don't need to talk about prerequisites: we get 100% reproducibiliity between our two systems without wondering about what things might be missing. If there are things that are missing from the Ubuntu image, the `Dockerfile` can be edited to add them to the `apt install` lines, etc.

If you have trouble with the storing to s3 thing, then we can take it from there, provided you let me know what files are needed.

You said:

> Both ./artifacts/bin/coreclr/Linux.x64.Checked and ./.dotnet are created by build.sh, so I guess you can just tar gzip both folders and upload them to S3.

The trick is we need to know which files are needed at runtime. I don't know anything about dotnet, but having two directories seems tricky - what is the `.dotnet` directory for? Where should it get installed to? What about the `./artifacts`  - does that contain the compilers themselves?

File: build/buildCLR.sh (hunk 1 above)
Comment (mattgodbolt) [reply build/buildCLR.sh:11]: And...apologies @chausner ! You weren't the original PR author 😊 - I'm so sorry. I didn't mean to be so snappy here, sorry. I'll definitely help it over the finish line once I understand what's going on...

**Comment by mattgodbolt** [comment]
I see: we have a PR by @nabeelomer which has had some sensible suggestions made by @chausner :)  I'm neutral on this; I don't have the bandwidth to take on a whole new language (plus the build scripts), so I'm really looking for a "champion" to take this on. I can definitely help with the things that I know how to do (upload things to s3, CE changes) but I don't know enough about the CLR to make any headway on the specifics.

**Comment by kunalspathak** [comment]
The right command should be `build.sh clr -c Release -runtimeconfiguration Checked`

So here is the proposal:

1. `build.sh clr+libs -c Release -runtimeconfiguration Checked`
2. `build.sh generatelayoutonly /p:LibrariesConfiguration=Release` <-- this populates `artifacts/tests/coreclr/Linux.x64.Checked/tests/Core_Root` that contains all the libraries,  `crossgen2` needed to disassemble the program
3. Archive the `Core_Root` folder
4. Archive `./.dotnet` folder so we can use it for doing `dotnet build <<user code>>`
5. << When user inputs the program>>, `dotnet build` happens and produces `Program.dll`.
6. Execute `dotnet path/to/Core_Root/crossgen2/crossgen2.dll -O -targetarch x64 -targetos linux -r:path/to/Core_Root/*.dll --codegenopt NgenDiffableDasm=1 --codegenopt NgenDisasm=<MethodName>Program.dll`

Now, for `<Method_Name>` if we specify `*`, it will show the disassembly of all the methods that will be NGened as part of the compiling the user program (which the user don't want to see). One option would be to let user specify the method name they are interested in see the disassembly of in the options menu (see screenshot) and then we can use that method name to set in `COMPlus_NgenDisasm` environment.

image

We could also let user specify `COMPlus_NgenDump=<Method_Name> to spit on Jit dump that clrjit produces for a method.

7. Bonus, user can specify `--targetos windows --targetarch arm64` and that can help them see the Arm64 code produced by .NET for the same method.

**Comment by kunalspathak** [comment]
Thanks @mattgodbolt - I am happy to help. I will set some time for myself to work on this.

**Pull request closed without merging**